### PR TITLE
Avoid log(0) in KL divergence

### PR DESCRIPTION
**Pull request by bz-e**
…denominator and added a test case

### Describe your change:
Fixes #12233 
Added type NONE to make it pass type checking, and added a small constant to the kullback_leibler_divergence method to fix the bug of numerator and denominator being 0, and also added a test case.

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #12233 ".

### CODE REVIEW

**Review by imSanko — CHANGES_REQUESTED** [review]
what do you actually want to change ?

**Comment by Harmanaya** [comment]
@cclauss Could you please review this PR when you get a chance?

**Comment by kevin1kevin1k** [comment]
@bz-e I think we do not intend to change the default behavior when all y_true are nonzero.
A better way might be to mask out all the zero entries and only sum them.

**Comment by bz-e** [comment]
> @bz-e I think we do not intend to change the default behavior when all y_true are nonzero. A better way might be to mask out all the zero entries and only sum them.

I think this is the lowest time complexity solution to the issue #12233 .

**Comment by kevin1kevin1k** [comment]
> I think this is the lowest time complexity solution to the issue #12233 .

I meant you could do something like the following (need not be identical, just for demo), and the complexity stays linear.
Also IMHO since this repo is more for educational purpose than for practical/production usage, correctness is more favorable compared to efficiency

```python
    mask = y_true != 0
    y_true_filtered = y_true[mask]
    y_pred_filtered = y_pred[mask]
    kl_loss = y_true_filtered * np.log(y_true_filtered / y_pred_filtered)

```

**Comment by bz-e** [comment]
> I meant you could do something like the following (need not be identical, just for demo), and the complexity stays linear. Also IMHO since this repo is more for educational purpose than for practical/production usage, correctness is more favorable compared to efficiency
> 
> ```python
>     mask = y_true != 0
>     y_true_filtered = y_true[mask]
>     y_pred_filtered = y_pred[mask]
>     kl_loss = y_true_filtered * np.log(y_true_filtered / y_pred_filtered)
> ```

For educational purposes I think you are right, this is the method with the least changes.
